### ISSUE TITLE
More layout options + other ideas

### ISSUE THREAD

**Issue by doup** [pojome/elementor#575]
Hi there, first of all, thank you for the great work and making it available as open source!
I'm really happy I've just found this project. 😄 

I've tried Elementor during this afternoon and overall I like how it works… but I miss some options. I've prefered to keep all the ideas in this issue, since I don't now if the codebase allows to add this features and also seems that you already have enough open issues. 

This whislist is from a developers perspective and needs. In my use case, to try to have design cohesion I tend to limit what the wordpress user can do; and prefer to have total control of the styling via SASS/CSS. I'm also a big fan of CSS frameworks (Foundation/Bootstrap).

I know that you still are working on the docs, so maybe some of these are already available with the current API. Most of these can be breaking changes… but since you're on version `0.8.1` I have to ask, hehe. 😅 

I'm specially interested in 1, 2 & 3 and **I can contribute if the codebase allows these changes**.

Thank you!
## 1. New section types

I would like to have more complex section elements. I can think of two right now:
- Boxed: https://doup.github.io/munger/section/boxed.html
- Image to the Edge: https://doup.github.io/munger/section/edge-image.html

The boxed section can be done right now, but it's too complex for an end user (Section > 1 column > content; add padding/class + background to the 100% column). Plus, maybe I want to limit the available options for the section (e.g: just choose between 3 bg colors), or have a fixed padding that is consistent between the pages.

I can't think of any way of doing the **image to the edge** section right now. As you can see the content spans 8 columns (this would be user controlled) and the image takes the remaining space to the edge of the window. So it's a mix of between full-width & boxed.
## 2. Control over row HTML generation & 12cols based layouts

It would be nice to have 12 column based layouts, maybe it could be achieved by snapping the current percent based columns every 8.33% . After that it would be nice to output a Foundation or Bootstrap compatible columns HTML code.

Related: #535, #291.
## 3. Disable styling options

I found that right now it's OK how you handle the low level CSS styling and there is a good separation between content & styling. But, anyway it would be cool to have control over which attributes can be changed.

I suppose that this can be achieved through the Elements Control API. The same way, it would be nice to be able to hide some controls (eg. 'Columns Gap' in the Section) and give them a default value.

Reordering controls and changing them between tabs would be a cool addition.
## 4. Responsive column ordering/width

Right now column widths can be changed for `small` breakpoint. I would be nice to be able to change width also at `medium` breakpoint. Column ordering would be really nice also, although it's quite difficult to make it work with an UI. I think that Visual Composer has this resolved.
## 5. Widgets with SASS/BEM styling

This would be a great extra but probably has a really easy workaround: implement my own widgets. I've been working lately on an idea of making a set of –what you call– widgets (accordion, heading, button…) that can be restyled using SASS and follow [BEM naming convention](http://getbem.com/).

So, by default there can be an already compiled CSS for the "casual" user. And then, for the developers there could be some SASS files that can be integrated with our own SASS files and can be configured via SASS `$variables`. For extra styling the widget CSS class names could be named following BEM convention. The primary idea behind this is to avoid CSS inheritance and specificity problems (please, no more `!important`).

Really poor WIP proof of concept here: https://github.com/doup/munger
And the components: https://doup.github.io/munger
## 6. Configuration via code

I don't how you handle this, but as a developer I found it really valuable to be able to change every option of the plugin via code and save it to GIT. So that the configuration is forced by the theme.

**Comment by joshmarom** [comment]
Hi @doup.
Firstly, I apologize for taking too long to reply.
Secondly, I would like to thank you for taking the time to write this most insightful feedback and even more so for offering your active contribution.

I might not be able to address all the issues you brought up but we are certainly going to refer to this post as we go along developing.

Before I address your questions, I would also like to point out that we are currently in the process of developing our developer API, at the same time we are also developing an array of new widgets and features for our upcoming PRO version. This means that some of the flexibility you are looking for will be enabled once the API is ready. I also means it will be easier for us to collaborate with contributors once these phases are complete and the system is more stable.

**Section Types** - Section is a basic structure/layout element. It should have enough flexibility to allow a wide range of designs, but at the same time, not suggest any specific design.
Both "Section Types" you asked for can technically be achieved in Elementor (though It might require a greater level of experience and familiarity). For this purpose we intend to provide a selection of pre-built templates in our Library, to assist users with creating more complex designs.

**Control over row HTML generation** - Our API will include filters that will allow developers to customize the markup.

**Disable styling options** - Adding and removing widgets and controls will be supported by the API.

**Responsive column ordering/width** - We are going to need to look into it. I don't have a prediction about that at the moment.

**Widgets with SASS/BEM styling** - The concept sounds quite interesting. I would love to discuss these ideas with you in more detail. Anyway, as you can imagine, developing addon widgets will become pretty simple after the API will be introduced.

**Configuration via code** - We are considering adding this feature as a part of our API.

**Comment by doup** [comment]
Thanks for the reply, I'm short on time right now but once you publish the developers API I'll give it another look and try to solve the problems I found. Meanwhile I'll keep working (slowly) in a SASS/BEM enabled widget collection… I'm also trying to work out a theme framework on top of Timber (for which Elementor would be a nice addition) but this is an internal project for now that it'll need few months until it matures.

**Comment by doup** [comment]
Hi! I've seen that you've launched the PRO version, congratulations!

… but the blog post also reminded me about this issue... 😇 

You mention that the developer APIs are coming (and you've released a [hello-world example](https://github.com/pojome/elementor-hello-world)); I just wanted to know if you have an approximate roadmap for the API and if it's going to be possible to work out the main points I raised in the original post.

I'm specially looking forward for: `1. New section types` and `2. Control over row HTML generation & 12cols based layouts`. Is there any code example for this two issues? Or, can you point me to the PHP classes/methods that I should check to solve this issues?

Thank you for all the work!

**Comment by joshmarom** [comment]
Check out our API docs:
https://github.com/pojome/elementor/tree/develop/docs
